### PR TITLE
ion-java-cli issue found from testing ion-test-driver

### PR DESCRIPTION
**Pull request by cheqianh**
### Description:
This branch is used for ion-test-driver's ion-java build logic. I'll use this PR for testing and add more issues once I found them from testing ion-test-driver. 

### Procedure:

- [x] Testing/Finding issue
- [x] Fixing the issues that in the list
- [x] Ready for Merging
 

### Issues list
1. null container type value is not a SCALAR type, For example: null.list
**BEFORE**
{
 &nbsp;&nbsp; event_type: CONTAINER_START
  &nbsp;&nbsp; ion_type:..
  &nbsp;&nbsp; depth:..
}
{
  &nbsp;&nbsp; event_type: CONTAINER_END
  &nbsp;&nbsp; ion_type:..
  &nbsp;&nbsp; depth:..
}
**EXPECT:** something like below:
{
  &nbsp;&nbsp; event_type: SCALAR
  &nbsp;&nbsp; ion_type: list
  &nbsp;&nbsp; value_text: "null.list"
&nbsp;&nbsp; ......
}
2. Empty annotation and null annotation should performance same in this project but was treated differently.
3. Embedded stream doesn't update index so it never jump out of the while loop.
4. Compare mode should skip all the symbol_table events but process mode should keep them.
5. Need to add an annotation $ion_user_value to all IVM like symbols.
6. Can't recognize the file starting with the null.symbol (because of the logic checking if the file is start with $ion_event_stream)
7. Able to print incomplete event stream for `bad` vector.
8. Added some edge case check/ null check/ index check.
9. Compare mode should exit once it catches an exception.

### Test case: 
All test case from ion-tests

### Question/specific area wants to check:
1. https://github.com/cheqianh/ion-java/pull/9#issuecomment-688576528

### CODE REVIEW

**Comment by cheqianh** [comment]
above 2 commits fixed issue 1-3

**Comment by cheqianh** [comment]
below commit fixed issue 4

**Comment by cheqianh** [comment]
Above commit fixed issue 5

**Comment by cheqianh** [comment]
fixed issue 6

**Comment by cheqianh** [comment]
This commits just improved symbol_table check logic

**Comment by cheqianh** [comment]
Above 2 commits improved symbol_table and make sure embedded added annotation to CONTAINER_FIELD

**Comment by cheqianh** [comment]
Commit 74f82ca fix issue 7 which allows ion-java-cli print incomplete event stream when input file is invalid.

Steps I did for this commit:
1. Created a readContext class to store events that read from input. (line 1263)
2. Logic before is reading the whole input file then transfer the whole file to eventStream. Now is storing each event to readContext immediately after reading each entry. By this way, we will have an incomplete event stream in our readContext when we encounter errors. (start in 868, passes readContext to function ionStreamToEventStream)
3. Added an extra try...catch for reading input files to eventStream. Once we meet errors during the reading phase, we still can write the incomplete eventStream to output. (175 - 186)

**Comment by cheqianh** [comment]
Above 2 commits improved corner case checking. Wroked on issue 8.

**Comment by cheqianh** [comment]
Above 2 commits improved corner case checking and added some logic for error report type(READ/WRITE). Worked on issue 8.

**Comment by cheqianh** [comment]
Above 2 commits improved corner case checking and added some logic for error report type(READ/WRITE). Worked on issue 8.

**Comment by cheqianh** [comment]
Fixed issue 9

File: ion-java-cli/src/com/amazon/tools/cli/IonJavaCli.java [hunk 1]
```diff
@@ -1232,6 +1274,28 @@ private static BufferedOutputStream initOutputStream(CommandArgs args,
         }
     }
 
+    static class ReadContext {
+        private List<Event> eventStream;
```
Comment (tgregg) [inline ion-java-cli/src/com/amazon/tools/cli/IonJavaCli.java:1278]: Looks like this is only set by the constructor, so it should be `final`.

**Review by tgregg — APPROVED** [review]
One minor comment.
